Percolator/Percolator.tla

```diff
@@ -0,0 +1,417 @@
+                  \E w \in ws :
+                    /\ key_write' = [key_write EXCEPT ![k] = Append(@, w)]
+                    /\ UNCHANGED <<key_data>>
+
+\* Cleans up a stale lock when the client encounters one.
+cleanup(c) ==
+  LET
+    start_ts == client_ts[c].start_ts
+    primary == client_key[c].primary
+    secondary == client_key[c].secondary
+  IN
+    \/ /\ hasStaleLock(primary, start_ts)
+       /\ cleanupStaleLock(primary, start_ts)
+    \/ \E k \in secondary :
+         /\ hasStaleLock(k, start_ts)
+         /\ cleanupStaleLock(k, start_ts)
+
+\* Returns TRUE if there is no lock for key $k$, and no any newer write than
+\* $ts$.
+canLockKey(k, ts) ==
+  LET
+    writes == {w \in DOMAIN key_write[k] : key_write[k][w].commit_ts >= ts}
+  IN
+    /\ key_lock[k] = {}  \* no any lock for the key.
+    /\ writes = {}       \* no any newer write.
+
+\* Locks the key and places the data.
+lockKey(k, start_ts, primary) ==
+  /\ key_lock' = [key_lock EXCEPT ![k] = @ \union {[ts |-> start_ts, primary |-> primary]}]
+  /\ key_data' = [key_data EXCEPT ![k] = @ \union {start_ts}]
+  /\ UNCHANGED <<key_write>>
+
+\* Tries to lock primary key first, then the secondary key.
+lock(c) ==
+  LET
+    start_ts == client_ts[c].start_ts
+    primary == client_key[c].primary
+    secondary == client_key[c].secondary
+  IN
+    IF ~hasLockEQ(primary, start_ts)
+    THEN  \* primary key is not locked, lock it first.
+      /\ canLockKey(primary, start_ts)
+      /\ lockKey(primary, start_ts, primary)
+    ELSE  \* primary key has already been locked, choose a secondary key to lock.
+      \E k \in secondary :
+        /\ canLockKey(k, start_ts)
+        /\ lockKey(k, start_ts, primary)
+
+\* Returns TRUE if the client locks all keys.
+canCommit(c) ==
+  LET
+    start_ts == client_ts[c].start_ts
+    primary == client_key[c].primary
+    secondary == client_key[c].secondary
+  IN
+    /\ hasLockEQ(primary, start_ts)
+    /\ \A k \in secondary :
+        /\ hasLockEQ(k, start_ts)
+
+\* Commit the primary key.
+commitPrimary(c) ==
+  LET
+    start_ts == client_ts[c].start_ts
+    primary == client_key[c].primary
+  IN
+    /\ hasLockEQ(primary, start_ts)
+    /\ key_write' = [key_write EXCEPT ![primary] = Append(@, client_ts[c])]
+    /\ key_lock' = [key_lock EXCEPT ![primary] = @ \ {[ts |-> start_ts, primary |-> primary]}]
+    /\ UNCHANGED <<key_data>>
+
+Start(c) ==
+  /\ client_state[c] = "init"
+  /\ next_ts' = next_ts + 1
+  /\ client_state' = [client_state EXCEPT ![c] = "working"]
+  /\ client_ts' = [client_ts EXCEPT ![c].start_ts = next_ts']
+  /\ UNCHANGED <<key_lock, key_data, key_write, client_key>>
+
+Get(c) ==
+  /\ client_state[c] = "working"
+  /\ IF canPrewrite(c)
+       THEN
+         /\ client_state' = [client_state EXCEPT ![c] = "prewriting"]
+         /\ UNCHANGED <<next_ts, key_lock, key_data, key_write, client_ts, client_key>>
+       ELSE
+         /\ cleanup(c)
```

Comment:
if clean up failed, how can we enter aborted state?

Reply:
I prefer removing the aborted state for simplicity.

If we cannot clean up any locks, the corresponding client will just stay halted, i.e. it has no following states, stutters in the old state.

Reply:
This is not a full transaction flow, I suggest moving to the aborted state.

Reply:
No need to handle aborted in Next like https://github.com/tlaplus/Examples/blob/master/specifications/transaction_commit/TCommit.tla does. 

IMO, it is very strange that a transaction can keep the working state but not go to aborted.

Reply:
Discussed offline, agreed on lifting `Abort` as a separate action. Added more comments to describe its meaning.